lib/SILOptimizer/Mandatory/MoveOnlyChecker.cpp

```diff
@@ -40,6 +45,408 @@ static void diagnose(ASTContext &Context, SourceLoc loc, Diag<T...> diag,
   Context.Diags.diagnose(loc, diag, std::forward<U>(args)...);
 }
 
+//===----------------------------------------------------------------------===//
+//                    Copy of Borrowed Projection Checker
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+struct LivenessInfo {
+  SmallVector<SILBasicBlock *, 8> discoveredBlocks;
+  PrunedLiveness liveness;
+  PrunedLivenessBoundary livenessBoundary;
+
+  LivenessInfo() : discoveredBlocks(), liveness(&discoveredBlocks) {}
+
+  void clear() {
+    discoveredBlocks.clear();
+    liveness.clear();
+    livenessBoundary.clear();
+  }
+
+  void computeBoundary() { livenessBoundary.compute(liveness); }
+
+  ArrayRef<SILInstruction *> getLastUsers() const {
+    return livenessBoundary.lastUsers;
+  }
+};
+
+/// Once we have finished checking that a base owned value is not consumed
+/// unnecessarily, we need to look at all instances where the owned value was
+/// borrowed and prove that it was never copied in its borrowed form because the
+/// borrowed value must be consumed.
+///
+/// As part of this optimization, we flatten the borrowed lifetime and eliminate
+/// any copies that are only consumed by a destroy_value. If we find that the
+/// flattened borrowed value needs a copy to extend the lifetime of the value,
+/// we treat the lifetime extending use as requiring a consuming use.
+struct CopyOfBorrowedProjectionChecker {
+  SmallVector<SILInstruction *, 4> foundConsumesOfBorrowedSubValues;
+  SmallVector<SILInstruction *, 4> foundCopiesOfBorrowedSubValues;
+  SmallVector<SILInstruction *, 4> foundDestroysOfBorrowedSubValues;
+  SmallVector<SILInstruction *, 4> foundRecursiveBorrows;
```

Comment:
@atrick NOTE: I am not actually using all of these right now. With time though I am going to need to eliminate any temporary unneeded copies that I find during my def->use walks.